Store required patches in codestream object

vmezzela opened the pull request:
This is an initial reorganization of the code around `scan` and the way we store the patches in preparation for fixing #153 .

The final fix can be implemented only after #160 is merged since that PR drops the use of quilt to apply the patches, which makes it more easy to selectively apply them.

I took the chance to further refactor `scan`. What was done before is:

1. Retrieve the filtered codestreams
2. First loop on them to retrieve the patched kernels (using `get_patched_kernels()`)
3. Second loop on the filtered codestreams to check these whose kernel belong to the list of patched_kernels, otherwise add them either to working_cs or unaffected_cs (or skipping due to missing support for arch)

The unnecessary thing was that we were storing the `cs.kernel` in `patched_kernels` when filtering with `get_patched_kernels()`. And then afterward we were checking which codestreams had that kernel with `if cs.kernel in patched_kernels`. This has been simplified by just filtering the patched codestreams (rather than their kernel). With this, a whole loop has been dropped because unneeded.

Now the process is:

1. Retrieve the filtered codestreams
2. Loop over the filtered codestreams and filter it into affected_cs, unaffected_cs, patched_cs (in the same loop)

Avoiding the unneeded `patched_kernel.add(cs.kernel)` and then `cs` -> `cs.kernel` -> `if cs.kernel in patched_cs` -> `patched_cs.append(cs)` steps.

Besides that, the most notable part is that now the required patches are stored in the codestream object directly, so that we can just retrieve the patches from there on a per codestream basis rather than on a per product basis, effectively introducing the possibility to perform the patch analysis only on those that are needed. And same goes for the extraction where we will be able to apply only the needed ones.

This is yet to be tested. I am posting this PR in advance to help coordinate the work across different PRs.

vmezzela committed:
scan: simplify scan process

Signed-off-by: Vincenzo Mezzela <vincenzo.mezzela@suse.com>

vmezzela committed:
codestream: introduce get_base_branch function

Signed-off-by: Vincenzo Mezzela <vincenzo.mezzela@suse.com>

vmezzela committed:
klplib: patch: retrieve required patches from codestream object

Signed-off-by: Vincenzo Mezzela <vincenzo.mezzela@suse.com>

vmezzela committed:
scan: add FIXME comment

Signed-off-by: Vincenzo Mezzela <vincenzo.mezzela@suse.com>

fgyanz commented on klpbuild/plugins/setup.py:
```diff
@@ -146,21 +146,15 @@ def setup_codestreams(lp_name, data):
                                                     data["archs"],
                                                     utils.get_workdir(lp_name))
 
-    # Add new codestreams to the already existing list, skipping duplicates
-    old_patched_cs = get_codestreams_data('patched_cs')
-    new_patched_cs = natsorted(list(set(old_patched_cs + patched_cs)))
-
-    set_codestreams_data(upstream=upstream, patched_cs=new_patched_cs,
-                         cve=data['cve'])
```
Not fully onboard with the idea of not exporting the `patched_cs`. We need that information in order to properly build the cover letter here: https://github.com/SUSE/klp-build/pull/166/commits/fb2c4637d0cf8f567a8a805e19658a03ee91ee4d

vmezzela replied:
Oh, didn't know that! I will, adjust it :)

fgyanz commented on klpbuild/plugins/scan.py:
```diff
@@ -183,10 +169,37 @@ def scan(cve, conf, lp_filter, download, archs=utils.ARCHS, savedir=None):
+    logging.info("Filtering already patched codestreams...")
+    affected_codestreams = []   # Codestreams without all the patches
+    unaffected_codestreams = [] # Codestreams belonging to a non affected product
+    patched_codestreams = []    # Codestreams with all the patches
+
+    for cs in codestreams:
+        bc = cs.base_cs_name()
+        suse_patches = patches[bc]
+
+        if not suse_patches:
+            unaffected_codestreams.append(cs)
+            continue
+
+        for patch in suse_patches:
+            if not cs.has_patch(patch):
+                # Store the patched required by this codestream for future use
+                cs.add_required_patch(patch)
+
+        if cs.needs_patches():
+            affected_codestreams.append(cs)
+        else:
+            patched_codestreams.append(cs)
```
Missing some logs. Ideally, I'd try keeping the same ones as before.

vmezzela replied:
True, I dropped them and forgot to add it back :D 

Will fix :)

fgyanz reviewed (commented):
Very cool @vmezzela !! I like the changes, just some minor comments.

vmezzela commented:
@fgyanz comments addressed!

marcosps commented:
@vmezzela can we already review? :D

vmezzela commented:
Hi Marcos, I'll mark this as ready to be reviewed when I rebase it on the latest `devel`.

vmezzela committed:
codestream: add has_patch function

Signed-off-by: Vincenzo Mezzela <vincenzo.mezzela@suse.com>

vmezzela committed:
codestream: add reuired_patches field

Signed-off-by: Vincenzo Mezzela <vincenzo.mezzela@suse.com>

vmezzela committed:
klplib: ksrc: use base_cs_name()

Signed-off-by: Vincenzo Mezzela <vincenzo.mezzela@suse.com>

vmezzela committed:
scan: move get_patched_kernels from ksrc

Signed-off-by: Vincenzo Mezzela <vincenzo.mezzela@suse.com>

vmezzela committed:
scan: move cs_is_affected from ksrc

Signed-off-by: Vincenzo Mezzela <vincenzo.mezzela@suse.com>

vmezzela committed:
utils: introduce filter_codestreams_by_arch function

Signed-off-by: Vincenzo Mezzela <vincenzo.mezzela@suse.com>

vmezzela committed:
scan: drop useless check

In this way, if no patches are supplied, all the codestreams will go in
the unaffected_codestreams list

Signed-off-by: Vincenzo Mezzela <vincenzo.mezzela@suse.com>

vmezzela committed:
codestreams: introduce get_candidate_patches_dirs

Signed-off-by: Vincenzo Mezzela <vincenzo.mezzela@suse.com>

vmezzela committed:
extract: rework apply_all_patches

This is in preparation to make apply_all_patches aware of patches
already present on each codestreams so that it can skip those.

This patch also drops logging.debug("Patches dir %s doesnt exists",
pdir) as it's kind of redundant with the next logs.

Signed-off-by: Vincenzo Mezzela <vincenzo.mezzela@suse.com>

vmezzela committed:
extract: skip patches that have already been applied

Signed-off-by: Vincenzo Mezzela <vincenzo.mezzela@suse.com>

marcosps commented on klpbuild/plugins/scan.py:
```diff
@@ -187,52 +171,43 @@ def scan(cve, conf, lp_filter, download, archs=None, savedir=None):
 
-def get_patched_kernels(codestreams, patches):
+def filter_affected_codestreams(codestreams, patches):
+    # TODO: remove this
```
Remove this :P

vmezzela replied:
Is it removed in a following commit? :D

marcosps replied:
fair :)

marcosps reviewed (approved):
LGTM!

fgyanz commented on klpbuild/plugins/extract.py:
```diff
@@ -557,18 +557,31 @@ def apply_all_patches(lp_name, cs):
+            logging.info("\tDropping %s", patch_name)
+            continue
+
+        logging.info("\tApplying %s", patch_name)
```
Very much nitpicking, but I'd propose keeping the previous log message. It's a purely cosmetic change, but I think it keeps more consistent with klp-build logs format.

before:
```
[2025-11-18 22:05:52] 15.5u30:5.14.21-150500.55.121: Applying /home/fgonzalez/klp/livepatches/bsc1251984/fixes/15.5/01-wifi-rtw88-delete-timer-and-free-skb-queue-when-unlo.patch...
[2025-11-18 22:05:53] Patches dir /home/fgonzalez/klp/livepatches/bsc1251984/fixes/15.5u31 doesnt exists
[2025-11-18 22:05:53] Applying patches on 15.5u31(5.14.21-150500.55.124) from /home/fgonzalez/klp/livepatches/bsc1251984/fixes/15.5
[2025-11-18 22:05:53] 15.5u31:5.14.21-150500.55.124: Applying /home/fgonzalez/klp/livepatches/bsc1251984/fixes/15.5/01-wifi-rtw88-delete-timer-and-free-skb-queue-when-unlo.patch...
```

now: 
```
Applying patches on 15.5u29(5.14.21-150500.55.116) from /home/fgonzalez/klp/livepatches/bsc1251984/fixes/15.5
	Applying wifi-rtw88-delete-timer-and-free-skb-queue-when-unlo.patch
Applying patches on 15.5u30(5.14.21-150500.55.121) from /home/fgonzalez/klp/livepatches/bsc1251984/fixes/15.5
	Applying wifi-rtw88-delete-timer-and-free-skb-queue-when-unlo.patch
Applying patches on 15.5u31(5.14.21-150500.55.124) from /home/fgonzalez/klp/livepatches/bsc1251984/fixes/15.5
	Applying wifi-rtw88-delete-timer-and-free-skb-queue-when-unlo.patch
```

fgyanz reviewed (approved):
Reviewed and tested. I proposed a small cosmetic change but by no means mandatory. Thanks a lot for the changes, LGTM!! :partying_face:
